plugin/pkg/managedseed/validator/admission.go

```diff
@@ -0,0 +1,373 @@
+				return err
+			}
+			spec.DNS.Provider = dnsProvider
+		}
+		// TODO If spec.DNS.Provider is not nil should we check if the user specified exactly the DNS provider we found above?
+		// I assume not, since the shoot may have multiple DNS providers
+
+		if spec.Ingress.Domain == "" {
+			spec.Ingress.Domain = ingressDomain
+		} else if spec.Ingress.Domain != ingressDomain {
+			return apierrors.NewBadRequest(fmt.Sprintf("seed ingress domain must match shoot DNS domain %s", ingressDomain))
+		}
+	} else {
+		if spec.DNS.IngressDomain == nil || *spec.DNS.IngressDomain == "" {
+			spec.DNS.IngressDomain = &ingressDomain
+		} else if *spec.DNS.IngressDomain != ingressDomain {
+			return apierrors.NewBadRequest(fmt.Sprintf("seed ingress domain must match shoot DNS domain %s", ingressDomain))
+		}
+	}
+
+	// Initialize and validate networks
+	if spec.Networks.Nodes == nil {
+		spec.Networks.Nodes = shoot.Spec.Networking.Nodes
+	} else if shoot.Spec.Networking.Nodes != nil && *spec.Networks.Nodes != *shoot.Spec.Networking.Nodes {
+		return apierrors.NewBadRequest(fmt.Sprintf("seed nodes CIDR must match shoot nodes CIDR %s", *shoot.Spec.Networking.Nodes))
+	}
+	if spec.Networks.Pods == "" && shoot.Spec.Networking.Pods != nil {
+		spec.Networks.Pods = *shoot.Spec.Networking.Pods
+	} else if spec.Networks.Pods != "" && shoot.Spec.Networking.Pods != nil && spec.Networks.Pods != *shoot.Spec.Networking.Pods {
+		return apierrors.NewBadRequest(fmt.Sprintf("seed pods CIDR must match shoot pods CIDR %s", *shoot.Spec.Networking.Pods))
+	}
+	if spec.Networks.Services == "" && shoot.Spec.Networking.Services != nil {
+		spec.Networks.Services = *shoot.Spec.Networking.Services
+	} else if spec.Networks.Services != "" && shoot.Spec.Networking.Services != nil && spec.Networks.Services != *shoot.Spec.Networking.Services {
+		return apierrors.NewBadRequest(fmt.Sprintf("seed services CIDR must match shoot services CIDR %s", *shoot.Spec.Networking.Services))
+	}
+
+	// Initialize and validate provider
+	if spec.Provider.Type == "" {
+		spec.Provider.Type = shoot.Spec.Provider.Type
+	} else if spec.Provider.Type != shoot.Spec.Provider.Type {
+		return apierrors.NewBadRequest(fmt.Sprintf("seed provider type must match shoot provider type %s", shoot.Spec.Provider.Type))
+	}
+	if spec.Provider.Region == "" {
+		spec.Provider.Region = shoot.Spec.Region
+	} else if spec.Provider.Region != shoot.Spec.Region {
+		return apierrors.NewBadRequest(fmt.Sprintf("seed provider region must match shoot region %s", shoot.Spec.Region))
+	}
+
+	return nil
+}
+
+func (v *ValidateManagedSeed) getSeedDNSProvider(ctx context.Context, shoot *gardencore.Shoot) (*gardencore.SeedDNSProvider, error) {
+	var dnsProvider *gardencore.SeedDNSProvider
+	dnsProvider, err := v.getSeedDNSProviderForCustomDomain(ctx, shoot)
+	if err != nil {
+		return nil, err
+	}
+	if dnsProvider == nil {
+		dnsProvider, err = v.getSeedDNSProviderForDefaultDomain(ctx, shoot)
+		if err != nil {
+			return nil, err
+		}
+	}
```

Comment:
This "order"/code was added with #3131 and was on purpose like this: The settings for a custom domain are "easier" to compute (wrt network I/O) than the settings for a default domain (listing is needed).
Now that the code was moved to an admission plugin that already leverages caches (the implementation in #3131 does not necessarily because the `CachedRuntimeClient` feature gate is still alpha and disabled by default) it does not make much of a difference anymore, I'd assume. Still I don't see a strong reason to change it now. It anyways depends on a concrete landscape which portion of the shoots is using a custom domain. Default domains are completely optional.

Reply:
I copied this logic from the current `use-as-seed` implementation. No strong opinion either, if you think it's better I will change it.